### DISCUSSION TITLE
Custom UI - edit "navigation" in builtin dialogs

### DISCUSSION THREAD

**Discussion by Trigve** [oleg-shilo/wixsharp#1602] [Q&A] [answered]
Hi,
I'm converting my project from pure wix to Wix#. I want to inject some custom dialogs in the installer. They should be in this order: Welcome -> License -> Features -> Custom Dialog 1 -> Custom Dialog 2 -> Progress -> Exit.

Now, Custom Dialog 1/2 should be shown only when some conditions are met (for instance, particular feature is being installed). So my question is, how do I modify the next/prev actions for the built-in dialogs (for instance when on Features page I want to somehow intercept the "next" button event and decide what should be shown next). The only option I see is custom Features dialog and there, update the code that handles the "next" button.

Thanks

**Comment by oleg-shilo** [comment]
You can create your setup from the Custom UI VS project template. This way you will have the source code for all dialogs.
When the dialog that you potentially want to skip is displayed (e.g. load event for WinForm) you can check some condition, in your case a feature is selected. If you need to skip this dialog you can just call `next_click` button handled to proceed to the next dialogue right away.

For checking the feature selection look in the FeaturesDialog.cs

**Reply by Trigve** [reply]
So I've tried the suggested solution, but it has 2 problems.

1. If I call the `next_clik` in the dialog load event, it also happens when the "back" button is pressed (and therefor user cannot go back). I didn't found any API that would tell me from which previous dialog the navigation happened so I could somehow filter it.
2. Even if the 1. was somehow fixed, the dialog is shown for the fraction of the second, which isn't optimal because user may think that there is some error in the installer because of skipping the pages.

Most straightforward solution would be to be able to add the show condition (for instance lambda) when creating the dialog sequence (like it's done with pure wix). Is there any possible solution to this?

**Reply by oleg-shilo** [reply]
Apologies. The answer I gave you was not the best approach.

I captured a direct approach in the Tips'n'Hints section.

You can find it here: https://github.com/oleg-shilo/wixsharp/wiki/Tips'n'Tricks#how-to-skip-a-dialog-based-on-a-runtime-condition

**Reply by Trigve** [reply]
Thanks for the reply.

Looking at the links you posted, isn't `shell.Dialogs.RemoveAt()` also destructive operations, meaning that dialog will be permanently removed in given session? Wouldn't it be appropriate to use `shell.GoTo<T>()`?

Anyway, I've simplified it a bit and I'm now showing the "customization" dialog always, and only set visible flags to the panels linked to enabled features. I think it's more practical than dialog manipulation in runtime.

**Reply by oleg-shilo** [reply, marked as answer]
>Looking at the links you posted, isn't shell.Dialogs.RemoveAt() also destructive operations, meaning that dialog will be permanently removed in given session? Wouldn't it be appropriate to use shell.GoTo<T>()?

Well, if a dialog is not needed anymore (this is the impression I got) removing is OK. 
And yes, `shell.GoTo<T>()` is a gentler approach.
